File: example/src/com/lisb/android/mediashrink/example/ExampleActivity.java
package com.lisb.android.mediashrink.example;

import java.io.File;
import java.io.IOException;

import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.provider.MediaStore;
import android.util.Log;
import android.view.ContextMenu;
import android.view.ContextMenu.ContextMenuInfo;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.TextView;
import android.widget.Toast;

import com.lisb.android.mediashrink.MediaShrink;
import com.lisb.android.mediashrink.example.R.id;
import com.lisb.android.mediashrink.example.R.layout;

public class ExampleActivity extends Activity implements OnClickListener {

	private static final String LOG_TAG = ExampleActivity.class.getSimpleName();

	private static final int MAX_WIDTH = 384;
	private static final int VIDEO_BITRATE = 500 * 1024;
	private static final int AUDIO_BITRATE = 128 * 1024;
	
	private static final String EXPORT_DIR = "exports";
	private static final String EXPORT_FILE = "video.mp4";
	private static final String SPREF_SELECTED_FILEPATH = "filepath";

	private static final int RCODE_SELECT_FROM_GALLARY = 1;
	private static final int RCODE_CAPTURE_VIDEO = 1;

	private View progress;
	private TextView txtSelectedVideoPath;
	private View btnStartReencoding;
	private View btnPlaySelectedVideo;
	private View btnPlayReencodedVideo;

	private AsyncTask<Void, Void, Exception> reencodeTask;
	private Uri selectedVideoPath;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(layout.activity_example);

		// wiring
		final View btnCaptureVideo = findViewById(id.btn_select_video);
		registerForContextMenu(btnCaptureVideo);
		btnCaptureVideo.setOnClickListener(this);

		btnPlaySelectedVideo = findViewById(id.btn_play_selected_video);
		btnPlaySelectedVideo.setOnClickListener(this);

		btnStartReencoding = findViewById(id.btn_start_reencoding);
		btnStartReencoding.setOnClickListener(this);

		btnPlayReencodedVideo = findViewById(id.btn_play_reencoded_video);
		btnPlayReencodedVideo.setOnClickListener(this);

		progress = findViewById(android.R.id.progress);
		txtSelectedVideoPath = (TextView) findViewById(id.txt_selected_video_path);

		final String filepath = getSharedPreferences().getString(
				SPREF_SELECTED_FILEPATH, null);
		if (filepath != null) {
			onFileSelected(Uri.parse(filepath));
		}

		if (getOutput().exists()) {
			btnPlayReencodedVideo.setEnabled(true);
		}

		if (!Environment.MEDIA_MOUNTED.equals(Environment
				.getExternalStorageState())) {
			// No External Storage
			txtSelectedVideoPath.setText("External storage is not mounted.");
			btnCaptureVideo.setEnabled(false);
			btnPlaySelectedVideo.setEnabled(false);
			btnStartReencoding.setEnabled(false);
			btnPlayReencodedVideo.setEnabled(false);

			Toast.makeText(this, "Please mount external storage.",
					Toast.LENGTH_LONG).show();
		}
	}

	@Override
	protected void onPause() {
		super.onPause();

		if (selectedVideoPath != null) {
			final Editor editor = getSharedPreferences().edit();
			editor.putString(SPREF_SELECTED_FILEPATH,
					selectedVideoPath.toString());
			editor.commit();
		}
	}

	@Override
	public void onCreateContextMenu(ContextMenu menu, View v,
			ContextMenuInfo menuInfo) {
		super.onCreateContextMenu(menu, v, menuInfo);
		switch (v.getId()) {
		case id.btn_select_video:
			getMenuInflater().inflate(R.menu.context_menu_select_video, menu);
		}
	}

	@Override
	public boolean onContextItemSelected(MenuItem item) {
		switch (item.getItemId()) {
		case id.select_from_gallary: {
			final Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
			intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true);
			intent.setType("video/*");
			if (intent.resolveActivity(getPackageManager()) != null) {
				startActivityForResult(intent, RCODE_SELECT_FROM_GALLARY);
			} else {
				Toast.makeText(this, "Activity Not Found.", Toast.LENGTH_SHORT)
						.show();
			}
			break;
		}
		case id.capture_video: {
			final Intent intent = new Intent(MediaStore.ACTION_VIDEO_CAPTURE);
			intent.putExtra(MediaStore.EXTRA_DURATION_LIMIT, 90);
			intent.putExtra(MediaStore.EXTRA_VIDEO_QUALITY, 1);
			if (intent.resolveActivity(getPackageManager()) != null) {
				startActivityForResult(intent, RCODE_CAPTURE_VIDEO);
			} else {
				Toast.makeText(this, "Activity Not Found.", Toast.LENGTH_SHORT)
						.show();
			}
			break;
		}
		}

		return super.onContextItemSelected(item);
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		switch (resultCode) {
		case RESULT_OK:
			onFileSelected(data.getData());
		}
	}

	private void onFileSelected(final Uri uri) {
		selectedVideoPath = uri;
		if (uri != null) {
			txtSelectedVideoPath.setText(uri.toString());
			btnPlaySelectedVideo.setEnabled(true);
			btnStartReencoding.setEnabled(true);
		} else {
			txtSelectedVideoPath.setText("");
			btnPlaySelectedVideo.setEnabled(false);
			btnStartReencoding.setEnabled(false);
		}
	}

	private SharedPreferences getSharedPreferences() {
		return PreferenceManager.getDefaultSharedPreferences(this);
	}

	private void reencode() {
		assert reencodeTask == null;

		progress.setVisibility(View.VISIBLE);
		reencodeTask = new AsyncTask<Void, Void, Exception>() {
			@Override
			protected Exception doInBackground(Void... params) {
				final MediaShrink shrink = MediaShrink
						.createMediaShrink(ExampleActivity.this);
				getOutput().getParentFile().mkdirs();
				shrink.setOutput(getOutput().getAbsolutePath());
				shrink.setMaxWidth(MAX_WIDTH);
				shrink.setVideoBitRate(VIDEO_BITRATE);
				shrink.setAudioBitRate(AUDIO_BITRATE);
				try {
					shrink.shrink(selectedVideoPath);
					return null;
				} catch (IOException e) {
					Log.e(LOG_TAG, "MediaShrink failed.", e);
					return e;
				}
			}

			@Override
			protected void onPostExecute(Exception result) {
				super.onPostExecute(result);
				reencodeTask = null;
				progress.setVisibility(View.GONE);

				if (result != null) {
					getOutput().delete();
					btnPlayReencodedVideo.setEnabled(false);
				} else {
					btnPlayReencodedVideo.setEnabled(true);
				}
			}
		};

		reencodeTask.execute();
	}

	private File getOutput() {
		return new File(getExternalFilesDir(EXPORT_DIR), EXPORT_FILE);
	}

	private void playVideo(Uri uri) {
		final Intent intent = new Intent(Intent.ACTION_VIEW);
		intent.setDataAndType(uri, "video/*");
		startActivity(intent);
	}

	// ===== OnClickListener ===== //

	@Override
	public void onClick(View v) {
		switch (v.getId()) {
		case id.btn_select_video:
			v.showContextMenu();
			break;
		case id.btn_start_reencoding:
			reencode();
			break;
		case id.btn_play_selected_video:
			playVideo(selectedVideoPath);
			break;
		case id.btn_play_reencoded_video:
			playVideo(Uri.fromFile(getOutput()));
			break;
		}

	}
}

File: example/src/main/java/com/lisb/android/mediashrink/example/ExampleActivity.kt
package com.lisb.android.mediashrink.example

import android.app.Activity
import android.content.ContentResolver
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.provider.MediaStore
import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.FileProvider
import androidx.lifecycle.lifecycleScope
import com.lisb.android.mediashrink.MediaShrinkQueue
import kotlinx.coroutines.launch
import timber.log.Timber
import java.io.File

class ExampleActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var progress: View
    private lateinit var txtSelectedVideoPath: TextView
    private lateinit var btnStartReencoding: View
    private lateinit var btnPlaySelectedVideo: View
    private lateinit var btnPlayReencodedVideo: View
    private lateinit var mediaShrinkQueue: MediaShrinkQueue
    private var selectedVideoUri: Uri? = null

    private val outputDir: File
        get() = File(filesDir, EXPORT_DIR)
    private val outputFile: File
        get() = File(outputDir, EXPORT_FILE)

    @Suppress("SetTextI18n")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_example)
        // wiring
        val btnCaptureVideo = findViewById<View>(R.id.btn_select_video)
        registerForContextMenu(btnCaptureVideo)
        btnCaptureVideo.setOnClickListener(this)
        btnPlaySelectedVideo = findViewById(R.id.btn_play_selected_video)
        btnPlaySelectedVideo.setOnClickListener(this)
        btnStartReencoding = findViewById(R.id.btn_start_reencoding)
        btnStartReencoding.setOnClickListener(this)
        btnPlayReencodedVideo = findViewById(R.id.btn_play_reencoded_video)
        btnPlayReencodedVideo.setOnClickListener(this)
        progress = findViewById(android.R.id.progress)
        txtSelectedVideoPath = findViewById(R.id.txt_selected_video_path)
        if (savedInstanceState != null) {
            val selectedUri = savedInstanceState.getParcelable<Uri>(SAVED_SELECTED_URI)
            selectedUri?.let { onFileSelected(it) }
        }
        if (outputFile.exists()) {
            btnPlayReencodedVideo.isEnabled = true
        }
        mediaShrinkQueue = MediaShrinkQueue(this.applicationContext, Handler(), filesDir,
                MAX_WIDTH, VIDEO_BITRATE, AUDIO_BITRATE, DURATION_LIMIT)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(SAVED_SELECTED_URI, selectedVideoUri)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        super.onCreateContextMenu(menu, v, menuInfo)
        if (v.id == R.id.btn_select_video) {
            menuInflater.inflate(R.menu.context_menu_select_video, menu)
        }
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.select_from_gallery -> {
                val intent = Intent(Intent.ACTION_GET_CONTENT)
                intent.putExtra(Intent.EXTRA_LOCAL_ONLY, true)
                intent.type = "video/*"
                if (intent.resolveActivity(packageManager) != null) {
                    startActivityForResult(intent, RCODE_SELECT_FROM_GALLERY)
                } else {
                    Toast.makeText(this, "Activity Not Found.", Toast.LENGTH_SHORT)
                            .show()
                }
            }
            R.id.capture_video -> {
                val intent = Intent(MediaStore.ACTION_VIDEO_CAPTURE)
                intent.putExtra(MediaStore.EXTRA_DURATION_LIMIT, DURATION_LIMIT)
                intent.putExtra(MediaStore.EXTRA_VIDEO_QUALITY, 1)
                if (intent.resolveActivity(packageManager) != null) {
                    startActivityForResult(intent, RCODE_CAPTURE_VIDEO)
                } else {
                    Toast.makeText(this, "Activity Not Found.", Toast.LENGTH_SHORT)
                            .show()
                }
            }
        }
        return super.onContextItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK) {
            onFileSelected(data!!.data)
        }
    }

    private fun onFileSelected(uri: Uri?) {
        selectedVideoUri = uri
        if (uri != null) {
            txtSelectedVideoPath.text = uri.toString()
            btnPlaySelectedVideo.isEnabled = true
            btnStartReencoding.isEnabled = true
        } else {
            txtSelectedVideoPath.text = ""
            btnPlaySelectedVideo.isEnabled = false
            btnStartReencoding.isEnabled = false
        }
    }

    private fun shrink() {
        progress.visibility = View.VISIBLE
        outputDir.mkdirs()

        lifecycleScope.launch {
            try {
                mediaShrinkQueue.queue(selectedVideoUri!!, Uri.fromFile(outputFile))
                progress.visibility = View.GONE
                btnPlayReencodedVideo.isEnabled = true
                Toast.makeText(this@ExampleActivity, "Success!", Toast.LENGTH_SHORT).show()
            } catch (t: Throwable) {
                Timber.tag(TAG).e(t, "Failed to shrink media.")
                progress.visibility = View.GONE
                btnPlayReencodedVideo.isEnabled = true
                outputFile.delete()
                var errorMessage = t.message
                if (errorMessage.isNullOrEmpty()) errorMessage = "Failed to sh≤rink."
                Toast.makeText(this@ExampleActivity, errorMessage, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun playVideo(uri: Uri?) {
        val intent = Intent(Intent.ACTION_VIEW)
        intent.setDataAndType(uri, "video/*")
        if (ContentResolver.SCHEME_CONTENT == uri!!.scheme) {
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
        startActivity(intent)
    }

    // ===== OnClickListener ===== //
    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_select_video -> v.showContextMenu()
            R.id.btn_start_reencoding -> shrink()
            R.id.btn_play_selected_video -> playVideo(selectedVideoUri)
            R.id.btn_play_reencoded_video -> playVideo(FileProvider.getUriForFile(this,
                    "com.lisb.android.mediashrink.example.fileprovider", outputFile))
        }
    }

    companion object {
        private const val TAG = "ExampleActivity"
        private const val MAX_WIDTH = 384
        private const val VIDEO_BITRATE = 500 * 1024
        private const val AUDIO_BITRATE = 128 * 1024
        private const val DURATION_LIMIT: Long = 90
        private const val EXPORT_DIR = "exports"
        private const val EXPORT_FILE = "video.mp4"
        private const val SAVED_SELECTED_URI = "selected_uri"
        private const val RCODE_CAPTURE_VIDEO = 1
        private const val RCODE_SELECT_FROM_GALLERY = 2
    }
}